### DISCUSSION TITLE
Is 'plugin marketplace' a thing?

### DISCUSSION THREAD

**Discussion by ice1000** [com-lihaoyi/mill#1416] [Q&A] [answered]
I'm thinking about creating an ANTLR4 plugin (or is it available already?) and I'm not sure how to do it. I have a project built with Gradle and I'm fed up with it🤣

**Comment by lefou** [comment, marked as answer]
Welcome to the mill ecosystem!

Mill already has some plugins. Some are developed as separate projects and some are located in the mill repository under `contrib`.

See https://com-lihaoyi.github.io/mill/mill/Thirdparty_Modules.html and https://com-lihaoyi.github.io/mill/mill/Contrib_Modules.html.

Each way has it's own pros and cons. Have a look at some of these, and you get an idea, what is needed to write your own pluginprojects. 

But to get started, it isn't even needed to create a dedicated project. The easiest way to get started is to just write your own module (e.g. `AntlrModule`) and place it next to your `build.sc` (e.g. `AntlrModule.sc`), and include it via `import $ivy.AntlrModule`. 

Once you are sattisfied with your module and want to share it, create a dedicated project, copy the `AntlrModule.sc` as `AntlrModule.scala` into the source tree and add the mill dependencies to the project. Please make sure to use `ivyCompileDeps` for any mill dependencies, though.
